File: leetcode/lc-us/cpp/47.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    int n;
    vector<bool> st;
    vector<int> path;
    vector<vector<int>> res;

    void dfs(vector<int> &nums, int u) {
        if (u == n) {
            res.push_back(path);
            return;
        }
        for (int i = 0; i < nums.size(); i++) {
            if (st[i] || i > 0 && nums[i] == nums[i - 1] && !st[i - 1]) continue;
            path[u] = nums[i];
            st[i] = true;
            dfs(nums, u + 1);
            st[i] = false;
        }
    }

public:
    vector<vector<int>> permuteUnique(vector<int> &nums) {
        n = (int) nums.size();
        st.resize(n);
        path.resize(n);
        sort(nums.begin(), nums.end());
        dfs(nums, 0);
        return res;
    }
};

File: leetcode/lc-us/py/47.py
from typing import List


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        n, res = len(nums), []
        path = [0] * n
        st = [False] * n
        nums.sort()

        def dfs(u: int) -> None:
            if u == n:
                res.append(path[:])
                return
            for i in range(n):
                if st[i] or i and nums[i - 1] == nums[i] and not st[i - 1]:
                    continue
                st[i] = True
                path[u] = nums[i]
                dfs(u + 1)
                st[i] = False

        dfs(0)
        return res
